File: exchange/converter.go
package exchange

import (
	"crypto-tray/config"
	"crypto-tray/providers"
)

// Converter handles currency conversion for price data
type Converter struct {
	fetcher       *Fetcher
	configManager *config.Manager
}

// NewConverter creates a new currency converter
func NewConverter(fetcher *Fetcher, configManager *config.Manager) *Converter {
	return &Converter{
		fetcher:       fetcher,
		configManager: configManager,
	}
}

// ConvertPrices applies exchange rate conversion to price data
func (c *Converter) ConvertPrices(data []*providers.PriceData) {
	cfg := c.configManager.Get()
	rates := c.fetcher.GetRates()
	rate := 1.0
	currency := cfg.DisplayCurrency

	if rates != nil {
		if r, ok := rates.Rates[cfg.DisplayCurrency]; ok {
			rate = r
		}
	}

	for dataIdx := range data {
		data[dataIdx].ConvertedPrice = data[dataIdx].Price * rate
		data[dataIdx].Currency = currency
	}
}

File: src-tauri/src/exchange/converter.rs
use std::sync::Arc;

use crate::config::ConfigManager;
use crate::providers::PriceData;

use super::fetcher::Fetcher;

pub struct Converter {
    fetcher: Arc<Fetcher>,
    config_manager: Arc<ConfigManager>,
}

impl Converter {
    pub fn new(fetcher: Arc<Fetcher>, config_manager: Arc<ConfigManager>) -> Self {
        Self {
            fetcher,
            config_manager,
        }
    }

    pub async fn convert_prices(&self, data: &mut [PriceData]) {
        let cfg = self.config_manager.get();
        let rates = self.fetcher.get_rates().await;
        let mut rate = 1.0;
        let currency = cfg.display_currency.clone();

        if let Some(ref exchange_rates) = rates {
            if let Some(&r) = exchange_rates.rates.get(&cfg.display_currency) {
                rate = r;
            }
        }

        for price_data in data.iter_mut() {
            price_data.converted_price = price_data.price * rate;
            price_data.currency = currency.clone();
        }
    }
}
